### ISSUE TITLE
Consider enforcing min_batch_size check in aggregator

### ISSUE THREAD

**Issue by wangshan** [ietf-wg-ppm/draft-ietf-ppm-dap#271]
`min_batch_size` as part of task parameters is distributed out of band by aggregators. At the moment there's no protocol level support to guarantee `min_batch_size` is not maliciously reduced by a dishonest leader. For use cases that work with central differential privacy, the `min_batch_size` is chosen based on an epsilon-dp guarantee over a certain batch. A dishonest leader can send a smaller `min_batch_size` to helper to break a strong dp guarantee.
 
One possible solution is to include parameters like `min_batch_size` (and other privacy parameters, for eg. epsilons in differential privacy) in `Extension`, which will become part of AAD therefore avoid tempering, and enforcing aggregators to check the out-of-band `min_batch_size` matches the ones included in every `Extension` field for the same task. Aggregator can also implement a one-off calculation on the task parameters to verify the `min_batch_size` is reasonable with other provided privacy parameters. For example, given a local differential privacy epsilon e_0, target central differential privacy epsilon e_c, and a min_batch_size N,  we can calculate whether e_c can be achieved  by aggregating e_0 over N reports, by using various shuffling amplification analysis.

**Comment by BranLwyd** [comment]
`min_batch_size` is distributed out-of-band, which means that the helper and leader are considered to both have the value configured separately (i.e. without a chance for one of the aggregators modify the value that is configured for the other aggregator).

`min_batch_size` is also evaluated separately and independently by each aggregator for each collect operation. The leader's check is specified in [Collection Initialization](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#name-collection-initialization) & the helper's check is specified in [Collection Aggregation](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#name-collection-aggregation). (In both cases, the relevant section refers to the [Validating Batch Parameters](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#name-validating-batch-parameters) section which specifies the `min_batch_size` check directly.)

Given the above, I believe a malicious leader cannot reduce `min_batch_size`, either at time of task initialization or at time of collection. (A malicious leader can certainly skip its own `min_batch_size` check, but as long as the helper does not collude, a small batch will still be caught by the helper's check.)

> Aggregator can also implement a one-off calculation on the task parameters to verify the `min_batch_size` is reasonable with other provided privacy parameters.

This is an interesting idea & worthy of consideration IMO -- AFAIK, the DAP spec currently does not speak to what values of `min_batch_size` are reasonable.

**Comment by wangshan** [comment]
I think the problem is DAP doesn't specify how task parameters are distributed out-of-band. If the task parameters is sent to helper from a malicious leader before task initialisation, then the helper would check against a wrong `min_batch_size` later.

Having said that, my above suggestion of checking against a `min_batch_size` in AAD also assumes client have the right task parameters.

On the other hand, I see there is an open issue in [Differential privacy](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#name-differential-privacy):
```
[OPEN ISSUE: While parameters configuring the differential privacy noise (like specific distributions / variance) can be agreed upon out of band by the aggregators and collector, there may be benefits to adding explicit protocol support by encoding them into task parameters.]
```
These DP parameters will face the same problem: how do we guarantee they are configured honestly.

I agree we should consider adding `min_batch_size` calculation from epsilon-dp parameters. There are well researched paper about calculating `min_batch_size` based on local DP from each client and a desired central DP on the batch. This _may_ need to be VDAF specific though.

**Comment by simon-friedberger** [comment]
Regarding the values of `min_batch_size`, there is an explicit statement in "[6.4 Batch Parameters](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#name-batch-parameters)":

> This document does not specify how to choose minimum batch sizes.

I think there also used to be a somewhat obvious statement that if `min_batch_size` is 1 there is no privacy but I cannot find it anymore.

I assume specification is avoided because there might be corner cases where low values make sense, e.g. some tasks might have a batch size of 2 because plausible deniability is enough. I would also like to see some guidance/examples on this but probably in the VDAF spec.

**Comment by cjpatton** [comment]
As @simon-friedberger points out, the DAP draft doesn't try to define a set of configuration parameters that are sufficient for all deployments. Indeed, this is something we expect to be deployment specific, particularly when DP is desired.

@wangshan it sounds like what you're after here is a means of opting out of a task if its configuration does not provide a sufficient level of privacy (for some definition of "privacy"). I think some part of this mechanism is inherently deployment-specific, but it might be useful to specify some protocol logic that is likely to be common to multiple deployments.

One question is whether this mechanism should be implemented during the "online" phase of the protocol, as you suggest (i.e., during upload, aggregation, or collection) or "offline" at configuration time. Offline seems preferable to me, given that this is always treated as a fatal error. However since we haven't yet specified how tasks are configured, it's not clear that we can define any behavior here.

Suppose each Aggregator implemented an endpoint, e.g., `POST /add_task`, to which the Collector posted task configurations. We could allow the aggregator to "abort" this request with a specific error, e.g., `insufficientPrivacy`. Would this be sufficient?

**Comment by tgeoghegan** [comment]
I don't think that it makes sense to have the `min_batch_size` set by the client and covered by the HPKE AAD. There's no way for the client to verify or enforce that either aggregator actually honored the value covered by the report AAD, so I don't think it would mitigate any of the currently possible attacks.

Further, I believe that in order to undermine the `min_batch_size` privacy parameter, you need both/all aggregators to collude -- it doesn't do leader any good to emit aggregate shares over a single report unless helper computes an aggregate share over that same report. If you admit colluding aggregators, then you're outside of DAP's threat model: if they're already colluding, then they might as well just send each other decrypted input shares.

I think the same applies to aggregators using the wrong epsilon on an aggregate share -- it won't help unless both aggregators subvert the protocol together.

>Suppose each Aggregator implemented an endpoint, e.g., `POST /add_task`, to which the Collector posted task configurations. We could allow the aggregator to "abort" this request with a specific error, e.g., `insufficientPrivacy`. Would this be sufficient?

I don't think we can specify an `/add_task` endpoint in DAP without opening the can of worms of bringing task parameter negotiation into the scope of the protocol.

**Comment by wangshan** [comment]
> @wangshan it sounds like what you're after here is a means of opting out of a task if its configuration does not provide a sufficient level of privacy (for some definition of "privacy"). I think some part of this mechanism is inherently deployment-specific, but it might be useful to specify some protocol logic that is likely to be common to multiple deployments.

Yes. I agree complete mitigation will likely be deployment-specific. 

> Further, I believe that in order to undermine the min_batch_size privacy parameter, you need both/all aggregators to collude

That's only the case after task initialization. But the parameters can be corrupted at task initialization stage, I imagine it's very common for the leader's organisation to drive the task, if helpers just receive task parameters from leader, then they are unable to verify if the parameters from leader can guarantee any privacy for the task.

**Comment by wangshan** [comment]
This issue essentially captures the transparency and parameter enforcement part of issue https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/290

**Comment by chris-wood** [comment]
> I don't think that it makes sense to have the min_batch_size set by the client and covered by the HPKE AAD. There's no way for the client to verify or enforce that either aggregator actually honored the value covered by the report AAD, so I don't think it would mitigate any of the currently possible attacks.

@tgeoghegan I think the threat model here necessarily assumes that at least one aggregator is honest. I think you're right that if both were dishonest, including `min_batch_size` in Report has no benefit. But if at least one is honest, then doesn't the client have assurance that at least one aggregator checked it and responded accordingly?

**Comment by simon-friedberger** [comment]
I think there are three issues being conflated here:

1. A malicious leader reducing `min_batch_size` which, it was assumed that leader, helpers and collector agreed on this value out of band so this should be impossible.
2. Aggregators actually enforcing `min_batch_size` (what the title says) which they always had to do.
3. Clients actually knowing what `min_batch_size` will be enforced which is the transparency part.

I think 3. is a great idea, 1. should probably be discussed in #290 and 2. is a misunderstanding.
